README.md
# Magic Tower
Magic Tower is a RPG adventure game that player can fight with some monsters
and escape the tower. (Updating...)
## Getting Started

### Prerequisites
Python 3.7 is needed to play this game. Python 3.7 to install Python.

### Installed packages for running the program
- platformio-ide-terminal

## Running the program
```python
python main.py
```

### Break down into end to end program
```python
updating
```

## Authors
- Qian Xiang

## Contributing
Pull requests are welcome. For major changes, please open an issue first to
discuss what you would like to change.

Please make sure to update tests as appropriate.

## License
There is not any license yet.

## Acknowledgments
- People who gave me some advises

## Tips for teacher
- If you are looking for the assignment about list, please open the inventory.py.
- If you are looking for the "RPG Nested Dictionaries", please go to the main
file.
- If you are looking for the redo assignment about list, please open the inventory.py and check the last line.


changelog.md
# RPG adventure game: Magic Tower Changelog
### Pre-released Version
v1.0
- created some lists about weapons in the inventory file
- added a title of the game in the main game file

v1.1
- fixed a spelling bug in the inventory file

v1.2
- create a introduction of characters

### Official release Version
(After the game can be run, all updated will be here.)

-
-


main.py
# Class: CS 30
# Period: 1
# Date created: 2020-02-07
# Date last modified: 2020-03-11
# Name: Qian Xiang
# Description: main.py in my text base adventure game that combine other file
# and run the game


# import python standard library
import sys
import time


# defined my printing speed
def print_slow(str):
    for char in str:
        time.sleep(.3)
        sys.stdout.write(char)
        sys.stdout.flush()

# print the sentence
print_slow('Welcome to the Magic Tower!')


# create a nested dictionary (a dictionary in a dictionary) for the characters
# number of character is use to count the # of character we have and help us
# to create other code easier
characters = {
             'Knight': {
                        'number': '#1', 'name': 'Sir William Marshal',
                        'background story': 'A knight...',
                        'beginning weapon': 'Wood Sword, Stone Sword, '
                        'Bone Knife, Stone Knife, Wood Knife, '
                        'or Stone Hatchet',
                        'ability': ' ',
                        'attack damage': '26',
                        'health': '250'},
             'Mage': {
                    'number': '#2', 'name': 'Alastor Moody',
                    'background story': 'A mage...',
                    'beginning weapon': 'Mechanical Umbrella - Thousands of '
                    'Organ',
                    'ability': 'Ignore enemy armor. (Alastor Moody is a great '
                    'mage from British. His power can destory other armor. He '
                    'is a good choice to defeat high armor monster, though his'
                    ' body is weak.)',
                    'attack damage': '30',
                    'health': '200'}
             }

# use some data from the above dictionary that it can be some variable to use.
# use for-loop to print out my statement about characters line by line.
for characters, characters_info in characters.items():
    number = characters_info["number"]
    print(f"\nCharacters {number}: {characters}")
    name = characters_info['name']
    weapon = characters_info['beginning weapon']
    ability = characters_info['ability']
    health = characters_info["health"]
    damage = characters_info["attack damage"]
    print(f"\tName: {name.title()}")
    print(f"\tAblility: {ability}")
    print(f"\tCharacter's weapon inventory: {weapon}")
    print(f"\t{characters} has {health} health point")
    print(f"\tBasic Attack Damage: {damage}")


# dictionary about location
locations = {'map #1': {
             'map name': "Symbol of Water--Nereus's wealth",
             'story': 'This is a old tower from a long time ago. '
             'We do not anything about it, except it create by Nereus.',
             'tasks': "\n\t"
             "1st: find the password"
             "\n\t"
             "2nd: beat the Nereus's incarnation, then find the door"
             "\n\t"
             "3rd: enter the correct password, then win the game"}}


# use some data from the above dictionary that it can be some variable to use.
# use for-loop to print out my statement about locations line by line.
for locations, locations_info in locations.items():
    map_name = locations_info["map name"]
    print(f"\nMap: {map_name}")
    story = locations_info['story']
    tasks = locations_info['tasks']
    print(f"\tbackground: {story}")
    print(f"\ttasks: {tasks}")


# dictionary about weapon(inventory)
# a dictionary of weapon in the game
weapons = {"Weapons' inventory": {
           'name': 'Stone Hatchet',
           'description': 'A wonderful heavy guy',
           'damage': '20'}}

# Print out the weapon from my dictionary
for weapons, weapons_info in weapons.items():
    W = weapons_info["name"]
    print(f"\n*Knight's weapons:\n{W}")
    W_description = weapons_info['description']
    W_damage = weapons_info['damage']
    print(f"\tdescription: It is Marshal's weapon. {W_description}")
    print(f"\tdamage: {W_damage}")


inventory.py
# Course: CS 30
# Period: 1
# Date created: 20/02/13
# Date last modified: 20/02/23
# Name: Qian Xiang
# Description: inventory for this game that store my weapon's name, data, and
# description
# Small tips: the inventory was build as lists. they will be changed to be
# dictionary after.

# import random
import random


# a list of common weapon in the game
commonWeapon = ['Wood Sword', 'Stone Sword', 'Bone Knife', 'Stone Knife',
                'Wood Knife', 'Stone Hatchet', 'Wood Dragger', 'Stone Dragger'
                'Broken Bow and Arrow - Dragon and Triger',
                'Wood Bow and Arrow']
# a list of rare weapon in the game
rareWeapon = ["Ice Wand - Frozen Ghost", "Wind Sword - Death Zone",
              "Fire Dagger - Fire Punishment", "Dark Dragger - Infinite",
              "Mechanical Umbrella - Thousands of Organ"]
# a list of epic weapon in the game
epicWeapon = ["Star Wand - Final Justice", "Death Stone - Dark Hallow",
              "Unknown Sword - Broken Sword from the Stone"]
# a list of legenary weapon in the game
legendaryWeapon = ["Chaos Mirror"]


# print the list "commonWeapon" to verify it performs properly
print(commonWeapon)
# print the list "rareWeapon" tol; verify it performs properly
print(rareWeapon)
# print the list "epicWeapon" to verify it performs properly
print(epicWeapon)
# print the list "legendaryWeapon" to verify it performs properly
print(legendaryWeapon)
# All the items in list "commonWeapon" are printed out in sentence
# using f-strings
print("You found a weapon in a chest.",
      f"You get a common weapon {random.choice(commonWeapon)}!")


# All the items in list "rareWeapon" are printed out in sentence
# using f-strings
print("You found a weapon on the monster's body.",
      f"You get a rare weapon {random.choice(rareWeapon)}!")


# All the items in list "epicWeapon" are printed out in sentence
# using f-strings
print("You killed the boss in this room.",
      f"You get a epic weapon {random.choice(epicWeapon)}!")


# All the items in list "epicWeapon" are printed out in sentence
# using f-strings
print("You successfully opened the treasure's door with three epic",
      "weapons and a historical reels.",
      f"You get a terrible weapon {legendaryWeapon}!")


# Use append() to add one new weapon to the list commonWeapon.
commonWeapon.append("Sword of Glory - Light")
# print the list after append.
print(commonWeapon)


# Use insert() to add one new weapon to the beginning of my list.
commonWeapon.insert(0, "Unknown Broken Sword with monsters' blood")
# print the list after insert.
print(commonWeapon)


# Use del() method to delete a weapoon in my list
del commonWeapon[0]
# print the list after delete a element.
print(commonWeapon)


# Use pop() to print a message to the player.
print("Sorry. The special weapon ", commonWeapon.pop(0), "was destoryed.")


# Use remove() to remove a epic weapon
epicWeapon.remove("Star Wand - Final Justice")
# print the list after remove a element.
print(epicWeapon)


# Store my list rareWeapon alphabetically
rareWeapon.sort()
# print the list after sort
print(rareWeapon)


# print a sorted list rareWeapon into reverse chronological order
print(sorted(rareWeapon, reverse=True))

# use reverse() to print the list commonWeapon into reverse chronological order
commonWeapon.reverse()
# print the list commonWeapon after reverse
print(commonWeapon)

# Use len() to count the number of my list commonWeapon
len(commonWeapon)

# Use index() to create a numerical list
numerical_list = []
for a in commonWeapon:
    p = commonWeapon.index(a)
    numerical_list.append(p)
print(numerical_list)

# Use index to find the weapons' posistion in my list
index = [1, 3, 7]
play_weapon = [commonWeapon[i] for i in index]
player_weapon = []
for i in index:
    player_weapon.append(commonWeapon[i])

print(player_weapon)

# print out a list of the individual items using loops
for i in commonWeapon:
    print(i)


pseudocode.py
# Course: CS 30
# Period: 1
# Date created: 20/01/30
# Date last modified: 20/01/30
# Name: Qian Xiang
# Description: psceudocodes for this game

# All importing file and python standard library:
# import sys
# import time

# Print out the title of game
# Two selections:
# First: Just print the sentence "Welcome to the Magic Tower"
# I would like to print it slowly that create a little bit visual
# effect
# import sys
# import time
#
# def print_slow(str):
#     for char in str:
#         time.sleep(.3)
#         sys.stdout.write(char)
#         sys.stdout.flush()
#
# print_slow('Welcome to the Magic Tower')
# Second: Print out the following code
# title =
# r"""                                _
#       __       __                  (_)
#      /  \     /  \     __ _   __ _  _   ___
#     /  _ \   / _  \   / _` | / _` || | / _/
#    /  / \ \_/ / \  \ | (_| || (_| || || (__
#   /  /   \   /   \  \ \__,_| \__, ||_| \__/
#  /  /     \_/     \  \         _/ |
# /__/               \__\       |___/
#           ____
# _________/   /
# |____   ____/
#     |  |
#     |  |  ___ __    _    __ ___   _ __
#     |  | / _ \\ \  /_\  / // _ \| '__/
#     |  || (_) |\ \// \\/ /|  __/| |
#     \__/ \___/  \_/   \_/  \___||_|
# """
# print(title)

# print introduce introduce text
# print("...")
# Set up the start menu
# choices on menu:
# *characters
# *map
# *quit
# *game information

# if the player select the *characters
# then print out some informations about characters
# character #1:
# name: knight Sir William Marshal
# background story: (construction)
# beginning weapon: Wood Sword, Stone Sword, Bone Knife, Stone Knife,
# Wood Knife, or Stone Hatchet.
# ability: (construction)
# attack damage: 26
# health: 250

# character #2:
# name: mage Alastor Moody
# background story: (construction)
# beginning weapon: Ice Wand - Frozen Ghost or
# Mechanical Umbrella - Thousands of Organ
# ability: Ignore enemy armor. (Alastor Moody is a great mage from British.
# His power can destory other armor. He is a good choice to defeat high armor
# monster, although his body is weak.)
# attack damage: 30
# health: 200
# after the player select the character. jump to the commonWeapon

# describe the background story of the map
# map name: symbol of water--Nereus's wealth
# story: (construction)
# task: 1st: find the password
#       2nd: beat the Nereus's incarnation, then find the door
#       3rd: enter the correct password, then win the game
# map craft:
#                    Top
#       password  | a chest   |   start point
#       shark     | sea triger|   black turtle
#       hagfish S | Nereus    |   evil fish king
#                    Bottom

# then enter the game(begin)

# if the player select the *map
# just print out the map above
# after select, the player can select the characters
# print out characters' information
# after select, enter the game(begin)

# print the plot
# plot: (construction)
# choice your movement: Upstair, Downstair, Left, or Right
# after choice, enter the new room
# print out the plot
# plot: (construction)
# print out monster's name
# print out monster's hp
# 4 choice to decide your reaction, attack(make damage to the boss),
# dodge(dodge the boss's attack and make a few damage to the boss, maybe fail),
# use your weapon ability, escape(back to the last room and print the
# movement chooses again)
# if you died in the room, you will have two choises: 1. replay the game in the
# room, then print out 4 choice again. 2. give up and fail in the game, then
# print out the statement for loser.)
# def fail(died):
#     for die in died:
#         time.sleep(.5)
#         sys.stdout.write(die)
#         sys.stdout.flush()
# fail('You was kill by the(). The tower rab your power and disolve your body.'
#     )
# If you win, you will get a random weapon and enter a plot that decide
# by the boss you beat.
# --> loop

# if you enter the passwor room, you will do the same process as other
# you just get a password after you win the boss.

# if you enter the Nereus' room
# --> loop
# after you beat the boss, you will have a oppturnity to enter the password
# then win the game. if the password is right, you will win the game. if the
# password is wrong, you will fail the game.

# --> win
# print out plot: (construction)

# --> fail
# print out plot: (construction)
# print out the statement for loser.


# boss information: (construction)
# 1st:
# 2st:
# 3st:
# 4th:
# 5th:
# 6th:
# 7th:

# inventory information:
# weapon: (construction)
'name': 'Wood Sword',
'description': 'A long sword that made by wood.',
'damage': '10',
'name': 'Stone Sword',
'description': 'A tough sword that made by stone',
'damage': '15',
'name': 'Bone Sword',
'description': 'A mysterious sword with snake blood',
'damage': '18',
'name': 'Stone Knife',
'description': 'A stone knife that look like fish',
'damage': '8',
'name': 'Wood Knife',
'description': 'A toy',
'damage': '6',
